### PR TITLE
Rearrange discussion of cancellation and rejection

### LINKED ISSUE #3895: Editorial Fixes for HTTP-QUIC
## Request Cancellation and Rejection 

1. it better be clarify  Implementation means Client/Server or both.

> When the client aborts reading a response, it indicates that this response is no longer of interest. **Implementations** SHOULD cancel requests by abruptly terminating any directions of a stream that are still open.

2. there are some of similar terms

- reset
- cancel
- reject
- abort
- (abruptly) terminate

it seems better to make them more clarify what each of them describes and how they can be done.

### PR DESCRIPTION
**Pull request by MikeBishop**
Fixes #3895

I still don't feel like this flows quite right, so suggestions appreciated.

### CODE REVIEW

File: draft-ietf-quic-http.md [hunk 1]
```diff
@@ -667,30 +667,35 @@ this limit are not guaranteed to be accepted.
...
-H3_REQUEST_CANCELLED.
-
-When a client resets a request with the error code H3_REQUEST_CANCELLED, a
-server MAY abruptly terminate the response using the error code
-H3_REQUEST_REJECTED if no processing was performed.  Clients MUST NOT use the
-H3_REQUEST_REJECTED error code, except when a server has requested closure of
-the request stream with this error code.
-
-If a stream is cancelled after receiving a complete response, the client MAY
+Once a request has been sent, it MAY be cancelled by either endpoint.  Clients
```
Comment (bencebeky) [inline draft-ietf-quic-http.md:670]: Does the request body (if exists) need to be sent completely before the client can cancel the request?  Do the request headers need to be sent completely?  Or may the stream be cancelled if only a the very first byte of the HEADERS frame has been delievered in a QUIC STREAM frame?

Would "Once a request stream is open, the request MAY be cancelled" be better?

File: draft-ietf-quic-http.md [hunk 2]
```diff
@@ -667,30 +667,35 @@ this limit are not guaranteed to be accepted.
...
+sending parts of streams and aborting reading on receiving parts of streams;
+see Section 2.4 of [QUIC-TRANSPORT].
+
+When the server cancels a request without performing any application processing,
+the request is considered "rejected."  The server SHOULD abort its response
+stream with the error code H3_REQUEST_REJECTED. In this context, "processed"
+means that some data from the stream was passed to some higher layer of software
+that might have taken some action as a result. The client can treat requests
+rejected by the server as though they had never been sent at all, thereby
+allowing them to be retried later on a new connection. Servers MUST NOT use the
```
Comment (bencebeky) [inline draft-ietf-quic-http.md:687]: Is there any scenario whereby the request could be retried on the same connection (in which case "on a new connection" would be too restrictive)?  BTW, why would a server reject a request but not close the connection?

File: draft-ietf-quic-http.md (hunk 1 above)
Comment (MikeBishop) [reply draft-ietf-quic-http.md:670]: Technically, the stream needn't have any bytes sent.

File: draft-ietf-quic-http.md (hunk 2 above)
Comment (MikeBishop) [reply draft-ietf-quic-http.md:687]: Good point.  In some ways, this is a holdover from HTTP/2, where it was possible for the client to exceed the server's stream limit, and REJECTED was a mechanism to close those streams without penalizing the client.  A server might still do that under load -- you can't take back stream credit, but you can reject incoming requests and not replace the stream credit.

Comment (bencebeky) [reply draft-ietf-quic-http.md]: That makes sense.  In that case, a client is allowed to retry the request later on the same connection.

File: draft-ietf-quic-http.md [hunk 3]
```diff
@@ -667,30 +667,36 @@ this limit are not guaranteed to be accepted.
...
+sending parts of streams and aborting reading on receiving parts of streams;
+see Section 2.4 of [QUIC-TRANSPORT].
+
+When the server cancels a request without performing any application processing,
+the request is considered "rejected."  The server SHOULD abort its response
+stream with the error code H3_REQUEST_REJECTED. In this context, "processed"
+means that some data from the stream was passed to some higher layer of software
+that might have taken some action as a result. The client can treat requests
+rejected by the server as though they had never been sent at all, thereby
+allowing them to be retried later. Servers MUST NOT use the H3_REQUEST_REJECTED
```
Comment (martinthomson) [inline draft-ietf-quic-http.md:688]:
```suggestion
allowing them to be retried later.

Servers MUST NOT use the H3_REQUEST_REJECTED
```

This seems like a good point to break this mega-paragraph.

**Pull request merged**